Fixing unexpected linebreak on single selects

eriknygren opened the pull request:
Fixes https://github.com/sagalbot/vue-select/issues/460

What
---
- Hiding the search input field if the component is in the single value
option.
- Making the search input field full width if no options are selected in
	either single or multi select mode.
- Shrinking it to width auto if there are selected entries in multi
	mode.

Why
---

Before:
![image](https://user-images.githubusercontent.com/5814512/38931077-f1fa4ee8-4309-11e8-9856-ce977a708bb7.png)

After:
![image](https://user-images.githubusercontent.com/5814512/38931094-fd802de6-4309-11e8-876c-b03a828c3cc0.png)

Reason for being broken:
![image](https://user-images.githubusercontent.com/5814512/38931113-0ddfe5aa-430a-11e8-9d37-c1c329759d52.png)

The component broke into two lines when selecting a a value in single
mode, because an empty, non-interactable input field was pushed down to
the next row if the selected entry had a long label.

PS not sure if I did something weird with the indentation in the spec file?

fractalocity commented:
It looks awesome! Nice fix @eriknygren

eriknygren commented:
Great, thanks guys! Let me know if you need anything more from me. Happy to help

BasConijn commented:
I think this is the same problem as #225.

eriknygren commented:
bump, would love to go back to pointing to the master branch on my project 😎 

@sagalbot any idea when you'll have some time to test this? anything I can do to help?

eriknygren commented:
@sagalbot can't see any regressions regarding `multiple` select mode, the input used to be `width: auto` during `multiple` select mode, and it remains so.

![may-14-2018 18-57-18](https://user-images.githubusercontent.com/5814512/40014560-c74131ae-57a8-11e8-9ae3-c79feb5d5564.gif)

Attached a quick video of how it interacts in multiple mode above, looks good to me but you know the component better than me!

aabraham88 commented:
@sagalbot first of all, I really appreciate this beautiful plugin, congrats for your work, it's very useful. I'm using it in an NGO project. Right now I'm experiencing this issue as the other ones commented above. I understand that is just a cosmetic issue, but It would be helpful if you guys can review this pull request and merge into master. Thanks for all. 👍

tursics commented:
bump ;) Merging this pull request will improve single selections a lot. Leave it as current state would be odd

eriknygren commented:
@tursics @aabraham88 @Lovatt as a temporary workaround you could always update your dependency to point to this PR until it's merged. That's what I'm doing.

So in your package.json go:
```
    "vue-select": "https://github.com/eriknygren/vue-select.git#3aec695",
```
Clearly we all prefer not being on a fork, but just thought I'd mention it.

Lovatt commented:
Thanks @eriknygren, i didn't know you could do that!  :relieved:

Lovatt commented:
@eriknygren i tried your branch in my project, i did indeed fix the formatting issues, but it looks like it breaks search

eriknygren commented:
> @eriknygren i tried your branch in my project, it did indeed fix the formatting issues, but it looks like it breaks search

@Lovatt hmm, in what way? I'm not convinced that's related to this change. But I'd like to confirm, thanks.

stevenharman commented:
Just discovered the same issue. Any hope for getting this merged 🔜?

stevenharman commented on src/components/Select.vue:
```diff
@@ -213,6 +214,16 @@
     cursor: pointer;
   }
+  .v-select input[type="search"].hidden {
+    display: none;
```
In my testing, `display: none` can mess up the vertical alignment of the clear button and the dropdown arrow. Perhaps we consider the following instead, as they preserve the height of the input element, and the vertical alignment of the other elements?

```css
width: 0px;
padding: 0;
```

eriknygren replied:
Good shout @stevenharman , will update

stevenharman commented:
Just discovered the same issue. Any hope for getting this merged 🔜?

For what it's worth, the following hack seems to work around the issue... for now. Maybe?

```css
.v-select.single .selected-tag + input[type="search"] {
  padding: 0;
  width: 0px !important;
}
```

stevenharman commented:
@sagalbot Sure. You can actually see it in action on the demo pages: https://codepen.io/sagalbot/pen/qxbPLY

Go there in Firefox/Chrome, and then turn on the responsive design dev tool. Shrink the viewport down to less than ~380px wide and you'll see the following:

<img width="253" alt="vue_select_-_validation_with_veevalidate" src="https://user-images.githubusercontent.com/48658/41479313-be55f3dc-7098-11e8-8514-63445dbabf41.png">

eriknygren commented:
@sagalbot just noticed an issue where the input was incorrectly hidden if the user re-opens the dropdown after already having a selected item. I added another commit to fix that, hopefully it’s fairly clear what’s going on. Let me know if you want me to squish the commits or something.

Also, added your suggestion @stevenharman.

https://github.com/sagalbot/vue-select/pull/512/commits/16b6bd7d09708b8637125571e46d28031478f83c

stevenharman commented:
Nice catch @sagalbot. Looking forward to this being merged and released! Thanks!

stevenharman commented:
Or, rather... nice catch, @eriknygren. And @sagalbot is that sufficient to reproduce, test the fix, and get this merged and released?

Thanks!

eriknygren commented:
bump @sagalbot

stevenharman commented:
@sagalbot @eriknygren The more I play with these changes, the more I'm certain that we should also be "hiding" the input for a multi-select, just as we do for single select. That is, remove the check for `!this.multiple` from the `hidden` class:

```javascript
inputClasses() {
  return {
    hidden: !this.isValueEmpty && !this.dropdownOpen,
    shrunk: this.multiple && !this.isValueEmpty,
    empty: this.isValueEmpty,
  }
}
```

Everything still seems to work as expected, with the added benefit of not adding a second, blank, line to multi-selects.

Possibly related - one way to fix this would be to use a flex-box layout for the selected options, input, loading, and clear buttons.

So two questions:

1. Can we also hide for multi-select?
2 @sagalbot would you be open to using a flex-box based layout for the internals?

eriknygren commented:
Thanks @stevenharman I'll add your suggestion and try it out later today

eriknygren commented:
@stevenharman I took your suggestion for a spin, it fixes the bug you've mentioned, and works in all but one scenario. Because the multiple select leaves the input on loosing focus the behaviour is a bit "out of the norm" when this occurs.

**Hopefully this gif explains what I mean:**


![multiple_bug](https://user-images.githubusercontent.com/5814512/42898967-8ed5a588-8abc-11e8-8378-bfb727bda9fa.gif)


I'd vote for leaving multi select to another pull request. I think this can be merged as-is and it'll add value. Then we can think of another solution for the multiple mode.

stevenharman commented:
@eriknygren Sounds good. I feel like much of this can be simplified and solved by #594, which I built with your changes in this PR in mind. Once this PR is merged, I'll revisit #594 and tweak it accordingly.

eriknygren commented:
Perfect, thanks @stevenharman. Let's get this show on the road 🛣

aabraham88 commented:
Excellent work guys! 🎊 Do you know when this will come available in a release? Thanks in advance.

eriknygren commented:
My pleasure, thanks for taking the time and writing this mega useful component!

aabraham88 commented:
Hi everyone. Sorry for bother you with this question, but how do I do to update the package with npm to have this fix? Thanks in advance.

kyrpas commented:
@aabraham88 I just managed to add this fix to my project. Here's what I did, I hope it works out for you too.

Since the changes have been merged to master, we could pull the master branch using `npm install --save sagalbot/vue-select#pull/512/head` but it still won't work because the changes are only in the src folder and when you're importing vue-select in your project it pulls the files from the dist folder. So you need to build the dist files again.

According to #306 if you want to rebuild the dist folder, you need to clone the repo in a separate folder. It won't work if you try to npm run install and build inside the vue-select folder in the node_modules of your project.

So, first, I cloned the project into a local directory: `git clone https://github.com/sagalbot/vue-select.git`
Then went into the folder and ran `npm install`. 
Finally I ran `npm run build`.

That built the project and re-created the dist folder using the latest code from `src`.

I then copied over the files from the new dist folder over the node_modules/vue-select/dist folder in my project and everything worked fine. 

This really doesn't feel like the right way to do it but until there's an official release this works fine for me.

aabraham88 commented:
Hi @kyrpas I will try your suggestion. Thanks!

eriknygren commented:
@aabraham88 @kyrpas Did you try what I previously suggested?
I rebuilt the dist on my master branch, so in your `package.json`:
```
"vue-select": "https://github.com/eriknygren/vue-select.git#3aec695",
```

With you current solution @kyrpas you'll loose the fix if you re-download the dependencies

kyrpas commented:
@eriknygren I did see your comment on June 4, then realised your fork is a few commits behind the original master branch and chose to rebuild from master. To be honest, none of the two solutions are ideal but since they're only temporary I'm ok with either one. I see your point though. Your suggestion is more robust. Any idea as to when to expect a new release? Is there anything we can do to help?

cor commented:
I'm also very curious on a release date of a version that includes this patch. When can we expect this? I'll use @eriknygren 's fix in the meantime.

aabraham88 commented:
Hi everyone. I had the opportunity to test this in my project, and I think I found a bug. Here is the gif video of the test. 
http://recordit.co/bUl3YnLj9Q
In my case, I'm using the plugin with Ajax to retrieve the data. The selected value appears there because is retrieved from the database when the component is initialized. The example show the scenario when the user wants to change the current value. Instead of clicking on the "X", I clicked on the Down Arrow, showing me that there isn't any option available. That's because the user didn't write down anything yet.
The expected behavior would be to let the user close the "empty list" and let him write another word replacing the current value.
The problem is that I can't close the dropdown, in order to see the "X" again, and there is no turning back in this state.
I copy here my component props configuration:
```
<v-select
        :options="dataCoordinadores"
        label="nombre"
        placeholder="Escribe el nombre o apellido"
        name="coordinador"
        id="coordinador"
        v-model="coordinadorSeleccionado"
        v-bind:disabled="this.readonly"
        :filterable=false
        @search="onSearch"
        :onChange=this.actualizarCoordinador()
>
</v-select>
```
I hope you can tell me if I'm doing something wrong or instead is an issue. I just want to say that this issue is not reproducible before the change introduced in this pull request.

Thanks!

aabraham88 commented:
@stevenharman no, in my case it didn't work. As you can see in the video, clicking on the  `^` doesn't close the list. I could check with another select input that has preloaded values, that this does not occur. It allows you to click on `^` and close the dropdown.

stevenharman commented:
Ahh, I just realized I also have #594 applied to my app. That outstanding PR is built atop this one, and has a much more versatile and robust layout scheme. I wonder... if you try that branch, do you still see the issue?

matneves commented:
Hi there, I'm still seen the unexpected linebreak on single selects.

![Captura de Tela 2019-05-21 às 20 51 58](https://user-images.githubusercontent.com/2514252/58138003-de251480-7c0a-11e9-8b73-0eb5048574d3.png)

I'm currently using vue-select v3.1.0 along with bootstrap v4.

matt-snider commented:
@matneves I had the same issue with vue-select v3.1.0 and bootstrap v4. This did it for me in a component that is using vue-select:
 
```vue
<template>
    <v-select
        class="my-select"
        //... other options
    />
</template>

<style scoped>
.my-select >>> input.vs__search {
    height: 0 !important;
    margin: 0 !important;
    padding: 0 !important;
}
</style>
```

Note, I'm using [vue-loader](https://vue-loader.vuejs.org) and the `>>>` is the [deep selector syntax](https://vue-loader.vuejs.org/guide/scoped-css.html#deep-selectors).

wilderdyd commented:
Hi, The same issue still persists in vue-select 3.10.3. I could solve this for my project with the next CSS properties:

`.v-select .vs__selected-options{
  flex-wrap: nowrap;
  white-space: nowrap;
  overflow: hidden;
}`
